syntax = "proto3";
package manager;

service Manager {
    rpc register(Void) returns (ShardInfo) {}
    rpc guild_count(Void) returns (GuildInfo) {}
    rpc checkin(ShardID) returns (Void) {}
    rpc publish_file(File) returns (Url) {}
    rpc all_guilds(Void) returns (stream Guild) {}
    rpc guild_update(stream Guild) returns (Void) {}
}

message Void {
    bool val = 1;
}

message ShardInfo {
    uint32 shard_id = 1;
    uint32 shard_count = 2;
}

message GuildInfo {
    uint32 guild_count = 1;
    uint32 user_count = 2;
}

message ShardID {
    uint32 shard_id = 1;
}

message File {
    string location = 1;
    string name = 2;
    string filetype = 3;
    bytes file = 4;
}

message Url {
    string Url = 1;
}

message Guild {
    fixed64 id = 1;
    string name = 2;
    string icon = 3;
    string splash = 4;
    fixed64 owner_id = 5;
    string region = 6;
    uint32 afk_timeout = 7;
    bool unavailable = 8;
    int32 max_members = 9;
    string banner = 10;
    string description = 11;
    int32 mfa_level = 12;
    uint32 premium_tier = 13;
    int32 premium_subscription_count = 14;
    string preferred_locale = 15;
    int32 member_count = 16;
    repeated string features = 17;
}
